k4FWCore/components/IOSvc.h

```diff
@@ -0,0 +1,112 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FWCORE_IOSVC_H
+#define FWCORE_IOSVC_H
+
+#include "Gaudi/Property.h"
+#include "GaudiKernel/IDataProviderSvc.h"
+#include "GaudiKernel/IHiveWhiteBoard.h"
+#include "GaudiKernel/IIncidentListener.h"
+#include "GaudiKernel/IIncidentSvc.h"
+#include "GaudiKernel/Service.h"
+
+#include "podio/ROOTReader.h"
+#include "podio/ROOTWriter.h"
+// #include "podio/RNTupleReader.h"
+// #include "podio/RNTupleWriter.h"
+
+#include "k4FWCore/KeepDropSwitch.h"
+
+#include "IIOSvc.h"
+
+#include <string>
+#include <type_traits>
+#include <vector>
+
+class IOSvc : public extends<Service, IIOSvc, IIncidentListener> {
+  using extends::extends;
+
+public:
+  // Gaudi doesn't run the destructor of the Services so we have to
+  // manually ask for the writer to be deleted so it will call finish()
+  ~IOSvc() override = default;
+
+  StatusCode initialize() override;
+  StatusCode finalize() override;
+
+  std::tuple<std::vector<std::shared_ptr<podio::CollectionBase>>, std::vector<std::string>, podio::Frame> next()
+      override;
+
+  std::shared_ptr<std::vector<std::string>> getCollectionNames() const override {
+    return std::make_shared<std::vector<std::string>>(m_collectionNames);
+  }
+
+  void setReadingCollectionNames(const std::vector<std::string>& names);
+  void setReadingFileNames(const std::vector<std::string>& names);
+
+protected:
+  Gaudi::Property<unsigned int> m_bufferNbEvents{
+      this, "BufferNbEvents", 20000,
+      "approximate size of the buffer used to prefetch rawbanks in terms of number of events. Default is 20000"};
+  Gaudi::Property<unsigned int> m_nbSkippedEvents{this, "NSkip", 0, "First event to process"};
+
+  Gaudi::Property<std::vector<std::string>> m_collectionNames{this, "CollectionNames", {}, "List of files to read"};
+  Gaudi::Property<std::vector<std::string>> m_readingFileNames{this, "input", {}, "List of files to read"};
+  Gaudi::Property<std::string>              m_writingFileName{this, "output", {}, "List of files to read"};
+  Gaudi::Property<std::vector<std::string>> m_outputCommands{
+      this, "outputCommands", {"keep *"}, "A set of commands to declare which collections to keep or drop."};
+  Gaudi::Property<std::string> m_inputType{this, "ioType", "ROOT", "Type of input file (ROOT, RNTuple)"};
+
+  std::mutex m_changeBufferLock;
+
+  KeepDropSwitch m_switch;
+
+  std::unique_ptr<podio::ROOTReader> m_reader{nullptr};
+  std::shared_ptr<podio::ROOTWriter> m_writer{nullptr};
+
+  std::shared_ptr<podio::ROOTWriter> getWriter() override {
+    if (!m_writer) {
+      m_writer = std::shared_ptr<podio::ROOTWriter>(new podio::ROOTWriter(m_writingFileName.value()));
+    }
+    return m_writer;
+  }
+
+  void deleteWriter() override {
+    if (m_writer) {
+      m_writer = nullptr;
+    }
+  }
+  void deleteReader() override {
+    if (m_reader) {
+      m_reader = nullptr;
+    }
+  }
```

Comment:
The intent is really to delete the Writer so I kept the name but used `reset()`, otherwise it can not write any output files. The comment for the destructor was for these two functions and I moved it.